File: ports/fortran/tictactoe.f90
! ***************************************************************************
!  TicTacToe - Enterprise Edition                                                   
!  Version: 1.0.0                                                                   
!  Date: 2027                                                                      
!  Description: A robust implementation of the classic Tic-Tac-Toe game            
!  utilizing advanced Fortran architecture patterns and enterprise-grade             
!  documentation standards.                                                         
! ***************************************************************************

program TicTacToe
    implicit none
    
    character(len=1), dimension(3,3) :: board
    character(len=1) :: current_player
    logical :: game_active
    integer :: row, col, io_status
    
    call initialize_game()
    
    do while (game_active)
        call display_board()
        call process_turn()
    end do
    
contains

    subroutine initialize_game()
        integer :: i, j
        do i = 1, 3
            do j = 1, 3
                board(i,j) = ' '
            end do
        end do
        current_player = 'X'
        game_active = .true.
    end subroutine initialize_game
    
    subroutine display_board()
        call system('clear')
        print *, ' ', board(1,1), ' | ', board(1,2), ' | ', board(1,3)
        print *, '---+---+---'
        print *, ' ', board(2,1), ' | ', board(2,2), ' | ', board(2,3)
        print *, '---+---+---'
        print *, ' ', board(3,1), ' | ', board(3,2), ' | ', board(3,3)
    end subroutine display_board
    
    logical function make_move(r, c)
        integer, intent(in) :: r, c
        make_move = .false.
        if (board(r,c) == ' ') then
            board(r,c) = current_player
            make_move = .true.
        end if
    end function make_move
    
    logical function check_win()
        integer :: i
        check_win = .false.
        
        ! Check horizontal
        do i = 1, 3
            if (board(i,1) == current_player .and. &
                board(i,2) == current_player .and. &
                board(i,3) == current_player) then
                check_win = .true.
                return
            end if
        end do
        
        ! Check vertical
        do i = 1, 3
            if (board(1,i) == current_player .and. &
                board(2,i) == current_player .and. &
                board(3,i) == current_player) then
                check_win = .true.
                return
            end if
        end do
        
        ! Check diagonals
        if (board(1,1) == current_player .and. &
            board(2,2) == current_player .and. &
            board(3,3) == current_player) then
            check_win = .true.
            return
        end if
        
        if (board(1,3) == current_player .and. &
            board(2,2) == current_player .and. &
            board(3,1) == current_player) then
            check_win = .true.
            return
        end if
    end function check_win
    
    subroutine switch_player()
        if (current_player == 'X') then
            current_player = 'O'
        else
            current_player = 'X'
        end if
    end subroutine switch_player
    
    subroutine process_turn()
        print *, "Player ", current_player, "'s turn"
        print *, "Enter row (1-3): "
        read(*,*, iostat=io_status) row
        print *, "Enter column (1-3): "
        read(*,*, iostat=io_status) col
        
        if (io_status == 0 .and. row >= 1 .and. row <= 3 .and. &
            col >= 1 .and. col <= 3) then
            if (make_move(row, col)) then
                if (check_win()) then
                    call display_board()
                    print *, "Player ", current_player, " wins!"
                    game_active = .false.
                else
                    call switch_player()
                end if
            else
                print *, "Invalid move! Cell already taken."
                call sleep(1)
            end if
        else
            print *, "Invalid input! Please enter numbers between 1 and 3."
            call sleep(1)
        end if
    end subroutine process_turn
    
end program TicTacToe

File: ports/c/tictactoe.c
/***************************************************************************
 * TicTacToe - Enterprise Edition                                                    
 * Version: 1.0.0                                                                    
 * Date: 2023                                                                       
 * Description: A robust implementation of the classic Tic-Tac-Toe game             
 * utilizing advanced C architecture patterns and enterprise-grade                    
 * documentation standards.                                                          
 ***************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>

#define BOARD_SIZE 3

typedef struct {
    char board[BOARD_SIZE][BOARD_SIZE];
    char current_player;
    bool game_active;
} Game;

void initialize_game(Game* game) {
    for (int i = 0; i < BOARD_SIZE; i++) {
        for (int j = 0; j < BOARD_SIZE; j++) {
            game->board[i][j] = ' ';
        }
    }
    game->current_player = 'X';
    game->game_active = true;
}

void display_board(const Game* game) {
    system("clear");
    for (int i = 0; i < BOARD_SIZE; i++) {
        printf(" %c | %c | %c\n", game->board[i][0], game->board[i][1], game->board[i][2]);
        if (i < BOARD_SIZE - 1) {
            printf("---+---+---\n");
        }
    }
}

bool make_move(Game* game, int row, int col) {
    if (row < 0 || row >= BOARD_SIZE || col < 0 || col >= BOARD_SIZE) {
        return false;
    }
    
    if (game->board[row][col] == ' ') {
        game->board[row][col] = game->current_player;
        return true;
    }
    return false;
}

bool check_win(const Game* game) {
    // Check rows
    for (int i = 0; i < BOARD_SIZE; i++) {
        if (game->board[i][0] == game->current_player &&
            game->board[i][1] == game->current_player &&
            game->board[i][2] == game->current_player) {
            return true;
        }
    }
    
    // Check columns
    for (int i = 0; i < BOARD_SIZE; i++) {
        if (game->board[0][i] == game->current_player &&
            game->board[1][i] == game->current_player &&
            game->board[2][i] == game->current_player) {
            return true;
        }
    }
    
    // Check diagonals
    if (game->board[0][0] == game->current_player &&
        game->board[1][1] == game->current_player &&
        game->board[2][2] == game->current_player) {
        return true;
    }
    
    if (game->board[0][2] == game->current_player &&
        game->board[1][1] == game->current_player &&
        game->board[2][0] == game->current_player) {
        return true;
    }
    
    return false;
}

void switch_player(Game* game) {
    game->current_player = (game->current_player == 'X') ? 'O' : 'X';
}

bool process_turn(Game* game) {
    int row, col;
    printf("Player %c's turn\n", game->current_player);
    printf("Enter row (1-3): ");
    if (scanf("%d", &row) != 1) {
        while (getchar() != '\n');
        return false;
    }
    printf("Enter column (1-3): ");
    if (scanf("%d", &col) != 1) {
        while (getchar() != '\n');
        return false;
    }
    
    row--; // Convert to 0-based indexing
    col--;
    
    if (!make_move(game, row, col)) {
        printf("Invalid move! Try again.\n");
        sleep(1);
        return false;
    }
    
    if (check_win(game)) {
        display_board(game);
        printf("Player %c wins!\n", game->current_player);
        game->game_active = false;
        return true;
    }
    
    switch_player(game);
    return true;
}

int main() {
    Game game;
    initialize_game(&game);
    
    while (game.game_active) {
        display_board(&game);
        if (!process_turn(&game)) {
            printf("Invalid input! Please enter numbers between 1 and 3.\n");
            sleep(1);
        }
    }
    
    return 0;
}
